aten/src/ATen/core/ATenDispatch.h

```diff
@@ -64,46 +65,59 @@ namespace detail {
     }
   };
 
+  // NB: take by const reference (Don't do universal forwarding here! You
+  // don't want to move into this function!)
   template <typename... Args>
   TensorTypeSet multi_dispatch_tensor_type_set(const Args&... args) {
     return MultiDispatchTensorTypeSet().apply(args...).ts;
   }
 }
 
+using FallbackBoxedFunction = void(const char* schema, torch::jit::Stack*);
+
+// Assume T is decayed
+template <typename T>
+using not_ok_to_box =
+  c10::guts::disjunction<
+    c10::guts::negation<
+      c10::guts::disjunction<
+        std::is_constructible<IValue, T>,
+        // TensorOptions are not directly constructible into IValue,
+        // but torch::jit::push knows how to handle them
+        std::is_same<TensorOptions, T>
+      >>
+#ifdef BUILD_NAMEDTENSOR
+    ,
+    // some constructors are templated (and therefore pass
+    // is_constructible), but do not actually work with all
+    // template arguments, so we must blacklist them explicitly
+    // TODO: The correct fix is to sfinae based on is_constructible of T
+    std::is_same<optional<ArrayRef<Dimname>>, T>
+#endif
+  >;
+
+template <class Result, class... Args>
+using supports_boxed_fallback =
+  c10::guts::negation<c10::guts::disjunction<
+    std::is_lvalue_reference<Result>,
+    not_ok_to_box<Result>,
```

Comment:
curious, why the double negatives?

Reply:
no good reason, just how I initially wrote the code. I think one (bad) reason for the negative is because I was trying to avoid saying `std::negation<std::is_same<...,...>>`